adr/0002-scripting-engine.md

```diff
@@ -0,0 +1,170 @@
+- `wait:` (i.e. [`ActionWait`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L37)), which supports two types of "status checks":
+  - `cluster:` (i.e. [`ActionWaitCluster`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L43)) -
+    perform status checks against K8s cluster resources
+  - `network:` (i.e. [`ActionWaitNetwork`](https://github.com/defenseunicorns/maru-runner/blob/main/src/types/actions.go#L51)) -
+    poll arbitrary HTTP/TCP endpoints for a given status code
+
+The `wait` action is really the only abstraction that is provided around
+shell scripting. As we seek to enhance or expand built-in capabilities,
+we have three high-level options for doing so:
+
+1. expand YAML-based DSL with more configuration options
+2. vendor additional "tools" and encourage `./zarf <tool> [...]` pattern
+3. **provide a cross-platform scripting engine (with builtins
+   for common tasks)**
+
+Below is an example of a common use case for HTTP status checks that
+is not readily solved by the existing `wait.network:` probe:
+
+```yaml
+tasks:
+  - description: SonarQube UI Status Check
+    maxRetries: 30
+    cmd: |
+      STATUS=$(curl -s 'https://sonarqube.uds.dev/api/system/status' | ./uds zarf tools yq '.status')
+      echo "SonarQube system status: ${STATUS}"
+      if [ $STATUS != "UP" ]; then
+        sleep 10
+        exit 1
+      fi
+```
+
+### Native Scripting Engine
+
+For the purposes of this proposal, we will scope the evaluation to libraries
+that can be embedded natively.
+
+Tools like [`zx`](https://google.github.io/zx/getting-started)
+(which provides a JS API on top of shell scripting) are interesting, but
+anything that is not written in Go would be difficult to integrate in a way
+that actually improves portability of user-defined scripts.
+
+[github.com/avelino/awesome-go](https://github.com/avelino/awesome-go#embeddable-scripting-languages)
+provides a fairly comprehensive list of embeddable scripting languages.
+There are some good options here, including:
+
+- [`starlark-go`](https://github.com/google/starlark-go), Go implementation
+  of Starlark: Python-like language with deterministic evaluation and hermetic
+  execution
+- [`starlet`](https://github.com/1set/starlet), which
+  enhances the `starlark` runtime with useful extensions like `http`
+- expression languages like [`expr`](https://github.com/expr-lang/expr),
+  [`cel`](https://github.com/google/cel-go) (which is used by Kubernetes[^4]),
+  and [`cue`](https://github.com/cue-lang/cue)
+- [`otto`](https://github.com/robertkrimen/otto), a JS parser/interpreter
+  written in Go
+
+These are all great, portable and secure options. However, with the exception
+of `starlet`, they are just language runtimes and lack rich APIs that we could
+expose directly to users. We would have to build these APIs ourselves from
+scratch.
```

Comment:
I think one analysis missing here would be "health of ecosystem" - I would summarize that after a little digging as:

`starlark-go` and `cel` are both Google backed and have a good history of releases - `starlet` has one release and is backed by just one contributor - `otto` is backed by a single developer (San Francisco, CA) but does have a decently sized community / release history - `risor` is also backed by a single developer (Sterling, VA) and does have a decently sized community / release history.

`cel` is also not turing complete which could be a big downside for more complicated things one might want to do.

Probably the only real choices based on that are `starlark-go`, `otto` and `risor` and I do agree that raw `starlark-go` is limiting and while we could create an std lib with `otto` there would be a lot more wiring (we'd likely need an embed fs with node modules in it)

Reply:
```suggestion
- [`starlark-go`](https://github.com/google/starlark-go), Go implementation
  of Starlark: Python-like language with deterministic evaluation and hermetic
  execution
- [`starlet`](https://github.com/1set/starlet), which
  enhances the `starlark` runtime with useful extensions like `http`
- expression languages like [`expr`](https://github.com/expr-lang/expr),
  [`cel`](https://github.com/google/cel-go) (which is used by Kubernetes[^4]),
  and [`cue`](https://github.com/cue-lang/cue)
- [`otto`](https://github.com/robertkrimen/otto), a JS parser/interpreter
  written in Go

These are all great, portable and secure options. However, with the exception
of `starlet`, they are just language runtimes and lack rich APIs that we could
expose directly to users. We would have to build these APIs ourselves from
scratch.

`starlark-go` and `cel` are both Google backed and have a good history of
releases, but `cel` is a configuration language and not turing-complete.
`starlet` has one release and is backed by just one contributor. `otto` is
backed by a single developer (San Francisco, CA) but does have a
decently sized community / release history.

For our purposes, the only reasonable choices are `starlark-go` and `otto`.
```

Reply:
@Racer159 does this capture it?